File: app/modules/GHCPAgentHubCore/Sources/GHCPAgentHub/Services/CLISessionMonitorService.swift
//
//  CLISessionMonitorService.swift
//  GHCPAgentHub
//
//  Service for monitoring active GitHub Copilot CLI sessions
//  Adapted from AgentHub for GitHub Copilot CLI
//

import Foundation
import Combine

// MARK: - CLISessionMonitorService

/// Service for monitoring active Copilot CLI sessions from ~/.copilot folder
public actor CLISessionMonitorService {

  // MARK: - Configuration

  private let copilotDataPath: String

  // MARK: - Publishers

  private nonisolated(unsafe) let repositoriesSubject = CurrentValueSubject<[SelectedRepository], Never>([])
  public nonisolated var repositoriesPublisher: AnyPublisher<[SelectedRepository], Never> {
    repositoriesSubject.eraseToAnyPublisher()
  }

  private nonisolated(unsafe) let allSessionsSubject = CurrentValueSubject<[CLISession], Never>([])
  public nonisolated var allSessionsPublisher: AnyPublisher<[CLISession], Never> {
    allSessionsSubject.eraseToAnyPublisher()
  }

  // MARK: - State

  private var selectedRepositories: [SelectedRepository] = []

  // MARK: - Initialization

  public init(copilotDataPath: String? = nil) {
    if let path = copilotDataPath {
      self.copilotDataPath = path
    } else {
      #if os(Windows)
      self.copilotDataPath = NSHomeDirectory() + "\\.copilot"
      #else
      self.copilotDataPath = NSHomeDirectory() + "/.copilot"
      #endif
    }
  }

  // MARK: - Repository Management

  /// Adds a repository to monitor
  @discardableResult
  public func addRepository(_ path: String) async -> SelectedRepository? {
    guard !selectedRepositories.contains(where: { $0.path == path }) else {
      return selectedRepositories.first { $0.path == path }
    }

    let worktrees = await detectWorktrees(at: path)

    let repository = SelectedRepository(
      path: path,
      worktrees: worktrees,
      isExpanded: true
    )

    selectedRepositories.append(repository)
    await refreshSessions()

    return repository
  }

  /// Removes a repository from monitoring
  public func removeRepository(_ path: String) async {
    selectedRepositories.removeAll { $0.path == path }
    repositoriesSubject.send(selectedRepositories)
  }

  /// Returns currently selected repositories
  public func getSelectedRepositories() -> [SelectedRepository] {
    selectedRepositories
  }

  /// Sets the list of selected repositories
  public func setSelectedRepositories(_ repositories: [SelectedRepository]) async {
    selectedRepositories = repositories
    await refreshSessions()
  }

  // MARK: - Session Scanning

  /// Refreshes sessions for all selected repositories
  public func refreshSessions() async {
    // Scan all sessions from session-state directory
    let allSessions = await scanAllSessions()

    // Filter sessions by selected repositories
    var updatedRepositories = selectedRepositories

    for repoIndex in updatedRepositories.indices {
      for worktreeIndex in updatedRepositories[repoIndex].worktrees.indices {
        let worktreePath = updatedRepositories[repoIndex].worktrees[worktreeIndex].path

        // Find sessions matching this worktree
        let matchingSessions = allSessions.filter { session in
          session.projectPath == worktreePath ||
          session.projectPath.hasPrefix(worktreePath + "/") ||
          session.projectPath.hasPrefix(worktreePath + "\\")
        }

        updatedRepositories[repoIndex].worktrees[worktreeIndex].sessions = matchingSessions.sorted {
          $0.lastActivityAt > $1.lastActivityAt
        }
      }
    }

    selectedRepositories = updatedRepositories
    repositoriesSubject.send(selectedRepositories)
    allSessionsSubject.send(allSessions)
  }

  /// Scan all sessions from the Copilot session-state directory
  public func scanAllSessions() async -> [CLISession] {
    let sessionStatePath = "\(copilotDataPath)/session-state"

    guard FileManager.default.fileExists(atPath: sessionStatePath) else {
      return []
    }

    var sessions: [CLISession] = []

    do {
      let sessionDirs = try FileManager.default.contentsOfDirectory(atPath: sessionStatePath)

      for sessionId in sessionDirs {
        // Skip non-UUID directories
        guard UUID(uuidString: sessionId) != nil else { continue }

        let sessionDir = "\(sessionStatePath)/\(sessionId)"
        let workspaceFile = "\(sessionDir)/workspace.yaml"
        let eventsFile = "\(sessionDir)/events.jsonl"

        // Read workspace metadata
        guard let metadata = readWorkspaceMetadata(at: workspaceFile) else { continue }

        // Check if active (events file modified recently)
        var isActive = false
        if let attrs = try? FileManager.default.attributesOfItem(atPath: eventsFile),
           let modDate = attrs[FileAttributeKey.modificationDate] as? Date {
          isActive = Date().timeIntervalSince(modDate) < 60
        }

        // Parse events for message count
        let parseResult = SessionEventsParser.parseSessionFile(at: eventsFile)

        // Get timestamps
        let createdAt = parseTimestamp(metadata.createdAt)
        let updatedAt = parseTimestamp(metadata.updatedAt)

        let session = CLISession(
          id: sessionId,
          projectPath: metadata.cwd,
          branchName: metadata.branch ?? "main",
          isWorktree: false,  // Will be determined by worktree detection
          lastActivityAt: updatedAt ?? parseResult.lastActivityAt ?? Date(),
          messageCount: parseResult.messageCount,
          isActive: isActive,
          firstMessage: parseResult.recentActivities.first(where: {
            if case .userMessage = $0.type { return true }
            return false
          })?.description,
          lastMessage: parseResult.recentActivities.last(where: {
            if case .userMessage = $0.type { return true }
            return false
          })?.description,
          summary: metadata.summary
        )

        sessions.append(session)
      }
    } catch {
      // Handle error silently
    }

    return sessions.sorted { $0.lastActivityAt > $1.lastActivityAt }
  }

  // MARK: - Worktree Detection

  private func detectWorktrees(at repoPath: String) async -> [WorktreeBranch] {
    // Run git worktree list
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
    process.arguments = ["worktree", "list", "--porcelain"]
    process.currentDirectoryURL = URL(fileURLWithPath: repoPath)

    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = Pipe()

    do {
      try process.run()
      process.waitUntilExit()

      let data = pipe.fileHandleForReading.readDataToEndOfFile()
      let output = String(data: data, encoding: .utf8) ?? ""

      var worktrees: [WorktreeBranch] = []
      var currentPath: String?
      var currentBranch: String?

      for line in output.components(separatedBy: .newlines) {
        if line.hasPrefix("worktree ") {
          currentPath = String(line.dropFirst("worktree ".count))
        } else if line.hasPrefix("branch refs/heads/") {
          currentBranch = String(line.dropFirst("branch refs/heads/".count))
        } else if line.isEmpty, let path = currentPath {
          let isMainWorktree = path == repoPath
          worktrees.append(WorktreeBranch(
            name: currentBranch ?? URL(fileURLWithPath: path).lastPathComponent,
            path: path,
            isWorktree: !isMainWorktree,
            sessions: []
          ))
          currentPath = nil
          currentBranch = nil
        }
      }

      // Handle last entry
      if let path = currentPath {
        let isMainWorktree = path == repoPath
        worktrees.append(WorktreeBranch(
          name: currentBranch ?? URL(fileURLWithPath: path).lastPathComponent,
          path: path,
          isWorktree: !isMainWorktree,
          sessions: []
        ))
      }

      if worktrees.isEmpty {
        // No worktrees, use main repo
        let branch = await getCurrentBranch(at: repoPath)
        return [WorktreeBranch(
          name: branch ?? "main",
          path: repoPath,
          isWorktree: false,
          sessions: []
        )]
      }

      return worktrees
    } catch {
      // Fallback to main repo
      let branch = await getCurrentBranch(at: repoPath)
      return [WorktreeBranch(
        name: branch ?? "main",
        path: repoPath,
        isWorktree: false,
        sessions: []
      )]
    }
  }

  private func getCurrentBranch(at path: String) async -> String? {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
    process.arguments = ["branch", "--show-current"]
    process.currentDirectoryURL = URL(fileURLWithPath: path)

    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = Pipe()

    do {
      try process.run()
      process.waitUntilExit()

      let data = pipe.fileHandleForReading.readDataToEndOfFile()
      return String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
    } catch {
      return nil
    }
  }

  // MARK: - Helpers

  private func readWorkspaceMetadata(at path: String) -> WorkspaceMetadata? {
    guard let data = FileManager.default.contents(atPath: path),
          let content = String(data: data, encoding: .utf8) else {
      return nil
    }

    var dict: [String: String] = [:]
    for line in content.components(separatedBy: .newlines) {
      let parts = line.split(separator: ":", maxSplits: 1)
      if parts.count == 2 {
        let key = String(parts[0]).trimmingCharacters(in: .whitespaces)
        let value = String(parts[1]).trimmingCharacters(in: .whitespaces)
        dict[key] = value
      }
    }

    guard let id = dict["id"], let cwd = dict["cwd"] else {
      return nil
    }

    return WorkspaceMetadata(
      id: id,
      cwd: cwd,
      gitRoot: dict["git_root"],
      repository: dict["repository"],
      branch: dict["branch"],
      summary: dict["summary"],
      summaryCount: Int(dict["summary_count"] ?? ""),
      createdAt: dict["created_at"],
      updatedAt: dict["updated_at"]
    )
  }

  private func parseTimestamp(_ string: String?) -> Date? {
    guard let string = string else { return nil }

    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

    if let date = formatter.date(from: string) {
      return date
    }

    formatter.formatOptions = [.withInternetDateTime]
    return formatter.date(from: string)
  }
}

File: electron-app/src/services/CLISessionMonitorService.ts
/**
 * CLISessionMonitorService
 * 
 * Service for monitoring GitHub Copilot CLI sessions.
 * Scans ~/.copilot/session-state/ for sessions.
 * Port of Swift CLISessionMonitorService for parity.
 */

import * as fs from 'fs';
import * as path from 'path';
import { EventEmitter } from 'events';
import { exec } from 'child_process';
import { promisify } from 'util';
import {
  CLISession,
  WorkspaceMetadata,
  WorktreeBranch,
  SelectedRepository,
} from '../models/types';
import { parseSessionFile } from './SessionEventsParser';
import { ConfigService, PersistedRepository } from './ConfigService';

const execAsync = promisify(exec);
const fsPromises = fs.promises;

/**
 * CLI Session Monitor Service
 */
export class CLISessionMonitorService extends EventEmitter {
  private copilotDataPath: string;
  private selectedRepositories: SelectedRepository[] = [];
  private configService: ConfigService | null = null;

  constructor(copilotDataPath?: string) {
    super();
    this.copilotDataPath = copilotDataPath ?? this.getDefaultCopilotPath();
  }

  /**
   * Set the config service for persistence
   */
  setConfigService(configService: ConfigService): void {
    this.configService = configService;
  }

  /**
   * Load saved repositories from config and restore them
   */
  async loadSavedRepositories(): Promise<void> {
    if (!this.configService) return;

    const savedRepos = this.configService.loadRepositories();
    for (const saved of savedRepos) {
      // Only restore if the path still exists
      if (fs.existsSync(saved.path)) {
        await this.addRepository(saved.path, false); // Don't save during load
        // Restore isExpanded state
        const repo = this.selectedRepositories.find(r => r.path === saved.path);
        if (repo) {
          repo.isExpanded = saved.isExpanded;
        }
      }
    }

    if (savedRepos.length > 0) {
      this.emit('repositoriesChanged', this.selectedRepositories);
    }
  }

  /**
   * Save current repositories to config
   */
  private saveRepositories(): void {
    if (!this.configService) return;

    const toSave: PersistedRepository[] = this.selectedRepositories.map(repo => ({
      path: repo.path,
      isExpanded: repo.isExpanded,
    }));

    this.configService.saveRepositories(toSave);
  }

  private getDefaultCopilotPath(): string {
    const home = process.env.HOME ?? process.env.USERPROFILE ?? '';
    return path.join(home, '.copilot');
  }

  /**
   * Get session state path
   */
  getSessionStatePath(): string {
    return path.join(this.copilotDataPath, 'session-state');
  }

  /**
   * Add a repository to monitor
   */
  async addRepository(repoPath: string, persist: boolean = true): Promise<SelectedRepository | undefined> {
    if (this.selectedRepositories.some(r => r.path === repoPath)) {
      return this.selectedRepositories.find(r => r.path === repoPath);
    }

    const worktrees = await this.detectWorktrees(repoPath);

    const repository: SelectedRepository = {
      path: repoPath,
      worktrees,
      isExpanded: true,
    };

    this.selectedRepositories.push(repository);
    await this.refreshSessions();

    if (persist) {
      this.saveRepositories();
    }

    return repository;
  }

  /**
   * Remove a repository from monitoring
   */
  removeRepository(repoPath: string): void {
    this.selectedRepositories = this.selectedRepositories.filter(r => r.path !== repoPath);
    this.saveRepositories();
    this.emit('repositoriesChanged', this.selectedRepositories);
  }

  /**
   * Get selected repositories
   */
  getSelectedRepositories(): SelectedRepository[] {
    return this.selectedRepositories;
  }

  /**
   * Set selected repositories
   */
  async setSelectedRepositories(repositories: SelectedRepository[]): Promise<void> {
    this.selectedRepositories = repositories;
    await this.refreshSessions();
  }

  /**
   * Refresh sessions for all repositories
   */
  async refreshSessions(): Promise<void> {
    const allSessions = await this.scanAllSessions();

    // Re-detect worktrees for each repository so new ones appear
    for (let repoIdx = 0; repoIdx < this.selectedRepositories.length; repoIdx++) {
      const repo = this.selectedRepositories[repoIdx];
      const updatedWorktrees = await this.detectWorktrees(repo.path);
      this.selectedRepositories[repoIdx].worktrees = updatedWorktrees;
    }

    // Filter sessions by selected repositories
    for (let repoIdx = 0; repoIdx < this.selectedRepositories.length; repoIdx++) {
      const repo = this.selectedRepositories[repoIdx];

      for (let wtIdx = 0; wtIdx < repo.worktrees.length; wtIdx++) {
        const worktree = repo.worktrees[wtIdx];

        // Find sessions matching this worktree
        const matchingSessions = allSessions.filter(session => {
          return session.projectPath === worktree.path ||
                 session.projectPath.startsWith(worktree.path + path.sep);
        });

        this.selectedRepositories[repoIdx].worktrees[wtIdx].sessions =
          matchingSessions.sort((a, b) => b.lastActivityAt.getTime() - a.lastActivityAt.getTime());
      }
    }

    this.emit('repositoriesChanged', this.selectedRepositories);
    this.emit('sessionsChanged', allSessions);
  }

  /**
   * Scan all sessions from session-state directory (async I/O)
   */
  async scanAllSessions(): Promise<CLISession[]> {
    const sessionStatePath = this.getSessionStatePath();

    if (!fs.existsSync(sessionStatePath)) {
      return [];
    }

    const sessions: CLISession[] = [];

    try {
      const sessionDirs = await fsPromises.readdir(sessionStatePath);

      await Promise.all(sessionDirs.map(async (sessionId) => {
        // Skip non-UUID directories
        if (!this.isValidUUID(sessionId)) return;

        const sessionDir = path.join(sessionStatePath, sessionId);
        const workspaceFile = path.join(sessionDir, 'workspace.yaml');
        const eventsFile = path.join(sessionDir, 'events.jsonl');

        // Read workspace metadata
        const metadata = this.readWorkspaceMetadata(workspaceFile);
        if (!metadata) return;

        // Check if active (async stat)
        let isActive = false;
        try {
          const stats = await fsPromises.stat(eventsFile);
          const secondsAgo = (Date.now() - stats.mtime.getTime()) / 1000;
          isActive = secondsAgo < 60;
        } catch {
          // File doesn't exist or can't be read
        }

        // parseSessionFile uses its own incremental cache — cheap when file unchanged
        const parseResult = parseSessionFile(eventsFile);

        // Find first/last user message
        const userMessages = parseResult.recentActivities.filter(
          a => a.activityType.type === 'userMessage'
        );
        const firstMessage = userMessages[0]?.description;

        // Parse timestamps
        const updatedAt = metadata.updatedAt ? new Date(metadata.updatedAt) : undefined;

        const session: CLISession = {
          id: sessionId,
          projectPath: metadata.cwd,
          branchName: metadata.branch ?? 'main',
          isWorktree: false,
          lastActivityAt: updatedAt ?? parseResult.lastActivityAt ?? new Date(),
          messageCount: parseResult.messageCount,
          isActive,
          firstMessage,
          lastMessage: undefined,
          summary: metadata.summary,
        };

        sessions.push(session);
      }));
    } catch (err) {
      console.error('Error scanning sessions:', err);
    }

    return sessions.sort((a, b) => b.lastActivityAt.getTime() - a.lastActivityAt.getTime());
  }

  /**
   * Detect worktrees for a repository
   */
  private async detectWorktrees(repoPath: string): Promise<WorktreeBranch[]> {
    try {
      const { stdout } = await execAsync('git worktree list --porcelain', {
        cwd: repoPath,
      });

      const worktrees: WorktreeBranch[] = [];
      let currentPath: string | undefined;
      let currentBranch: string | undefined;
      let isFirst = true;

      const pushEntry = () => {
        if (!currentPath) return;
        // First entry from `git worktree list` is always the main worktree
        const isMainWorktree = isFirst ||
          currentPath.toLowerCase() === repoPath.toLowerCase();
        isFirst = false;
        worktrees.push({
          name: currentBranch ?? path.basename(currentPath),
          path: currentPath,
          isWorktree: !isMainWorktree,
          sessions: [],
          isExpanded: true,
        });
        currentPath = undefined;
        currentBranch = undefined;
      };

      for (const line of stdout.split('\n')) {
        if (line.startsWith('worktree ')) {
          currentPath = line.substring('worktree '.length);
        } else if (line.startsWith('branch refs/heads/')) {
          currentBranch = line.substring('branch refs/heads/'.length);
        } else if (line === '' && currentPath) {
          pushEntry();
        }
      }

      // Handle last entry
      pushEntry();

      if (worktrees.length === 0) {
        // No worktrees, use main repo
        const branch = await this.getCurrentBranch(repoPath);
        return [{
          name: branch ?? 'main',
          path: repoPath,
          isWorktree: false,
          sessions: [],
          isExpanded: true,
        }];
      }

      return worktrees;
    } catch {
      // Fallback
      const branch = await this.getCurrentBranch(repoPath);
      return [{
        name: branch ?? 'main',
        path: repoPath,
        isWorktree: false,
        sessions: [],
        isExpanded: true,
      }];
    }
  }

  /**
   * Get current git branch
   */
  private async getCurrentBranch(repoPath: string): Promise<string | undefined> {
    try {
      const { stdout } = await execAsync('git branch --show-current', { cwd: repoPath });
      return stdout.trim() || undefined;
    } catch {
      return undefined;
    }
  }

  /**
   * Read workspace metadata from YAML file
   */
  private readWorkspaceMetadata(filePath: string): WorkspaceMetadata | undefined {
    if (!fs.existsSync(filePath)) return undefined;

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const dict: Record<string, string> = {};

      for (const line of content.split('\n')) {
        const colonIndex = line.indexOf(':');
        if (colonIndex > 0) {
          const key = line.substring(0, colonIndex).trim();
          const value = line.substring(colonIndex + 1).trim();
          dict[key] = value;
        }
      }

      if (!dict.id || !dict.cwd) return undefined;

      return {
        id: dict.id,
        cwd: dict.cwd,
        gitRoot: dict.git_root,
        repository: dict.repository,
        branch: dict.branch,
        summary: dict.summary,
        summaryCount: dict.summary_count ? parseInt(dict.summary_count, 10) : undefined,
        createdAt: dict.created_at,
        updatedAt: dict.updated_at,
      };
    } catch {
      return undefined;
    }
  }

  /**
   * Check if string is valid UUID
   */
  private isValidUUID(str: string): boolean {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    return uuidRegex.test(str);
  }
}
